apis/Google.Cloud.Spanner.Data/Google.Cloud.Spanner.Data/SpannerRetryHelpers.cs

```diff
@@ -0,0 +1,109 @@
+        {
+            if (retryOptions == null)
+            {
+                return await fn().ConfigureAwait(false);
+            }
+
+            DateTime start = clock.GetCurrentDateTimeUtc();
+            DateTime? overallDeadline = retryOptions.OverallTimeout.CalculateDeadline(clock);
+            IJitter jitter = new MinimumBoundJitter(retryOptions.DelayJitterMultiplier);
+            TimeSpan retryDelay = retryOptions.BackoffSettings.Delay;
+            while (true)
+            {
+                try
+                {
+                    var result = await fn().ConfigureAwait(false);
+                    return result;
+                }
+                catch (SpannerException e) when (retryOptions.ExceptionFilter(e))
+                {
+                    // If there's a recommended retry delay specified on the exception
+                    // we should at least wait for that.
+                    if (e.RecommendedRetryDelay.HasValue && e.RecommendedRetryDelay.Value > retryDelay)
+                    {
+                        retryDelay = e.RecommendedRetryDelay.Value;
+                    }
+                    TimeSpan actualDelay = jitter.GetDelay(retryDelay);
+                    DateTime expectedRetryTime = clock.GetCurrentDateTimeUtc() + actualDelay;
+                    if (expectedRetryTime > overallDeadline)
+                    {
+                        throw;
+                    }
+                    await scheduler.Delay(actualDelay, cancellationToken).ConfigureAwait(false);
+                    retryDelay = retryOptions.BackoffSettings.NextDelay(retryDelay);
+                }
+            }
+        }
+
+        private class MinimumBoundJitter : IJitter
+        {
+            private double _minDelayMultiplier;
+            private readonly object _lock = new object();
+
+            // See http://stackoverflow.com/questions/36376888 for why we don't have a thread-local RNG.
+            // We might create multiple instances of MinimumBoundJitter so this is static to make sure
+            // that we'll only have a single Random instance.
+            private static readonly Random s_random = new Random();
+
+            public MinimumBoundJitter(double minDelayMultiplier) => 
+                _minDelayMultiplier = GaxPreconditions.CheckArgumentRange(minDelayMultiplier, nameof(minDelayMultiplier), 1, double.MaxValue);
+
+            public TimeSpan GetDelay(TimeSpan minDelay)
+            {
+                if (minDelay <= TimeSpan.Zero || _minDelayMultiplier == 1)
+                {
+                    return minDelay;
+                }
+
+                double jitterMultiplier;
+                lock (_lock)
+                {
+                    jitterMultiplier = s_random.NextDouble();
+                }
+
+                var minDelayTicks = minDelay.Ticks;
```

Comment:
I don't think I fully understand what this code is meant to do, but I suspect it can be simplified. It'll be clearer with more comments about the purpose of the class/method.